pennylane/tape/qscript.py

```diff
@@ -395,41 +378,57 @@ def _update(self):
         """Update all internal metadata regarding processed operations and observables"""
         self._graph = None
         self._specs = None
-        self._update_circuit_info()  # Updates wires, num_wires; O(ops+obs)
-        self._update_par_info()  # Updates _par_info; O(ops+obs)
-
-        self._update_observables()  # Updates _obs_sharing_wires and _obs_sharing_wires_id
 
-    def _update_circuit_info(self):
-        """Update circuit metadata
+        try:
+            # Invalidate cached properties so they get recalculated
+            del self.wires
+            del self.par_info
+        except AttributeError:
+            pass
 
-        Sets:
-            wires (~.Wires): Wires
-            num_wires (int): Number of wires
-        """
-        self.wires = Wires.all_wires(dict.fromkeys(op.wires for op in self))
-        self.num_wires = len(self.wires)
+    @cached_property
+    def wires(self) -> Wires:
+        "Wires used in the quantum script process"
+        return Wires.all_wires(dict.fromkeys(op.wires for op in self))
 
-    def _update_par_info(self):
-        """Update the parameter information list. Each entry in the list with an operation and an index
-        into that operation's data.
+    @property
+    def num_wires(self) -> int:
+        """Number of wires in the quantum script process"""
+        return len(self.wires)
 
-        Sets:
-            _par_info (list): Parameter information
+    @cached_property
+    def par_info(self):
+        """list[dict[str, Operator or int]]: Parameter information.
+        Values are dictionaries containing the corresponding operation and operation parameter index.
         """
-        self._par_info = []
+        par_info = []
         for idx, op in enumerate(self.operations):
-            self._par_info.extend(
-                {"op": op, "op_idx": idx, "p_idx": i} for i, d in enumerate(op.data)
-            )
+            par_info.extend({"op": op, "op_idx": idx, "p_idx": i} for i, d in enumerate(op.data))
 
         n_ops = len(self.operations)
         for idx, m in enumerate(self.measurements):
             if m.obs is not None:
-                self._par_info.extend(
+                par_info.extend(
                     {"op": m.obs, "op_idx": idx + n_ops, "p_idx": i}
                     for i, d in enumerate(m.obs.data)
                 )
+        return par_info
+
+    @property
+    def obs_sharing_wires(self):
+        """list[~.Observable]: subset of the observables that share wires with another observable,
+        i.e., that do not have their own unique set of wires."""
+        if not self._obs_sharing_wires:
```

Comment:
What if there are no observables sharing wires? 

One potential option would be to distinguish `_obs_sharing_wires=None` (we haven't calculated it yet), and `_obs_sharing_wires=[]` (no observables are sharing wires).

Reply:
That is a fair point. I have added the corresponding check!